README.md
R package raincpc: Obtain and Analyze Rainfall data from the Climate Prediction Center (CPC)
========================================================

The Climate Prediction Center's daily rainfall data for the entire world, [CPC](www.cpc.ncep.noaa.gov), 1979 - present & 50-km resolution, is one of the few high quality and long-term observation-based rainfall products available for free. Data is available at CPC's [ftp site](ftp.cpc.ncep.noaa.gov/precip/CPC_UNI_PRCP/GAUGE_GLB/). However, it is a lot of data and there is no software to analyze and visualize the data.

Some issues with size/format of the CPC data:
* too many files (365/366 files per year * 34 years, separate folder for each year)
* each file has 360 rows and 720 columns
* file naming conventions have changed over time - one format prior to 2006 and couple of different formats afterwards
* file formats have changed over time - gzipped files prior to 2008 and plain binary files afterwards
* downloading multiple files simultaneously from the CPC ftp site, using wget, does not seem to work properly
* there is no software/code readily available to easily process/visualize the data

R package `raincpc`:
* Data for anytime period during 1979-present can be downloaded and processed
* Just two functions required: one to download the data (`cpc_get_rawdata`) and another to process the downloaded data (`cpc_read_rawdata`)
* Making spatial maps using the processed data is easy, via ggplot

Here are some examples.

Invoke required libraries.


```r
library(raincpc)

library(ggplot2)
library(reshape2)
```


Rainfall during Typhoon Yolanda/Haiyan, November 2013.


```r
# obtain data
cpc_get_rawdata(2013, 11, 7, 2013, 11, 7)
# process data
haiyan <- cpc_read_rawdata(2013, 11, 7)
```


Re-orient data for plotting.


```r
# function to flip a matrix upside down (change CPC orientation from S-N to
# N-S)
Fn_Flip_Matrix_Rows <- function(mat) {
    return(mat[nrow(mat):1, ])
}
# function to rotate a matrix 90 degress clockwise for plotting only used to
# counteract the 'image' function default behavior
Fn_Rotate_Matrix <- function(mat) {
    return(t(mat)[, nrow(mat):1])
}

haiyan <- Fn_Rotate_Matrix(Fn_Flip_Matrix_Rows(haiyan))
```


Global rainfall map.


```r
gfx_map <- ggplot(data = melt(haiyan))
gfx_map <- gfx_map + geom_raster(aes(Var1, Var2, fill = value))
gfx_map <- gfx_map + scale_fill_continuous(low = "blue", high = "red")
gfx_map <- gfx_map + theme(axis.text = element_blank(), axis.ticks = element_blank())
gfx_map <- gfx_map + labs(x = NULL, y = NULL, fill = "Rain (mm/day)")
gfx_map <- gfx_map + ggtitle("Rainfall during Typhoon Haiyan/Yolanda, Nov 7 2013")

png("gfx_haiyan.png", width = 800, height = 650)
plot(gfx_map)
garbage <- dev.off()
```


![figure 1] [fig1]

[fig1]: gfx_haiyan.png "figure 1" 


NEWS
Version 0.1
-----------

* Data download period limited to 1979 to 2012.

* Ways to make the data download faster needs to be explored.

* Future revision will include graphics functions to display the data.
    

R/cpc_get_rawdata.r
#' Download raw rainfall data from CPC for time period of interest
#'
#' This function obtains data from the CPC ftp site for any time period from
#' 1979/1/1 to the present day.
#' 
#' It is assumed that the ending year, month and day follow the beginning
#' year, month and day. Although CPC data is revised each day, a buffer
#' of two days is created - i.e., this function will only let you download
#' data till the day before yesterday! 
#' 
#' Output is either a ".gz" file (1979 - 2008) or a ".bin" file (2009 - 2013). 
#' 
#' @param begYr beginning year of the time period of interest, 1979 - 2012
#' @param begMo beginning month of the time period of interest, 1 - 12
#' @param begDay beginning day of the time period of interest, 1 - 28/29/30/31
#' @param endYr ending year of the time period of interest, 1979 -2012
#' @param endMo ending month of the time period of interest, 1 - 12
#' @param endDay ending day of the time period of interest, 1 - 28/29/30/31
#' @export
#' @examples
#' # CPC data for two days, Jul 11-12 2005
#' cpc_get_rawdata(2005, 7, 11, 2005, 7, 12)

cpc_get_rawdata <- function(begYr, begMo, begDay, endYr, endMo, endDay) {                            
  
  # check year validity
  if (!(begYr %in% seq(1979, 2013) & endYr %in% seq(1979, 2013))) {
    stop("Beginning and ending year should be between 1979 to 2013!")
  }
  # check dates validity
  check_dates <- try(seq(as.Date(paste(begYr, begMo, begDay, sep = "-")), 
                         as.Date(paste(endYr, endMo, endDay, sep = "-")), 
                         by = "day"), 
                     silent = TRUE)
  if (class(check_dates) == "try-error") {
    stop("Date range appears to be invalid!")
  }
  # check dates validity - ensure end year, month and day are before present date
  check_dates <- try(seq(as.Date(paste(endYr, endMo, endDay, sep = "-")),
                         Sys.Date() - 2, 
                         by = "day"), 
                     silent = TRUE)
  if (class(check_dates) == "try-error") {
    stop("End date should be two days before the present day!")
  }
    
  # url and file prefixes
  urlHead  <- "ftp://ftp.cpc.ncep.noaa.gov/precip/CPC_UNI_PRCP/GAUGE_GLB/"
  fileHead <- "PRCP_CU_GAUGE_V1.0GLB_0.50deg.lnx."
  
  for (eachYr in c(begYr:endYr)) {
    # generate dates of all the days in the format yyyy-mm-dd
    allDates <- seq(as.Date(paste(eachYr, begMo, begDay, sep = "-")), 
                    as.Date(paste(eachYr, endMo, endDay, sep = "-")), 
                    by = "day")
    # date string used in the filenames below
    dateStr  <- paste0(substr(allDates, 1, 4), 
                       substr(allDates, 6, 7), 
                       substr(allDates, 9, 10))
    yrStr    <- substr(allDates, 1, 4)
    
    # identify file names and extensions
    if (eachYr %in% seq(1979, 2005)) {
      urlTag  <- "V1.0/"
      fileTag <- ".gz"
    } else if (eachYr %in% c(2006)) {
      urlTag  <- "RT/"
      fileTag <- "RT.gz"
    } else if (eachYr %in% c(2007, 2008)) {
      urlTag  <- "RT/"
      fileTag <- ".RT.gz"
    } else { #if (eachYr %in% seq(2009, 2013)) {
      urlTag  <- "RT/"
      fileTag <- ".RT"
    } 
    
    # download
    for (eachDay in 1:length(allDates)) {
      # construct url
      fileUrl <- paste0(urlHead, urlTag, yrStr[eachDay], "/", fileHead, 
                        dateStr[eachDay], fileTag)
      
      # out file name; gzipped file prior to 2008 otherwise binary
      outFile <- ifelse(eachYr <= 2008, 
                        paste0("raw_", dateStr[eachDay], ".gz"), 
                        paste0("raw_", dateStr[eachDay], ".bin")) 
      
      # download file only if it doesnt exist
      # internet connection could be "intermittent" - could be due to the CPC server?
      # hence files are not sometimes downloaded; hence the quieted while loop below
      # max tries limited to 20 to avoid infinite loops
      if (!file.exists(outFile)) {
        fileError <- TRUE
        max_tries <- 0
        while (fileError & max_tries < 20) {
          max_tries <- max_tries + 1
          if (eachYr <= 2008) {
            fileStatus <- try(download.file(url = fileUrl, 
                                            destfile = outFile, 
                                            quiet = TRUE), 
                              silent = TRUE)
          } else {
            fileStatus <- try(download.file(url = fileUrl, 
                                            destfile = outFile, 
                                            mode = "wb", 
                                            quiet = TRUE), 
                              silent = TRUE)    
          }
          fileError  <- ifelse(class(fileStatus) == "try-error", TRUE, FALSE)
        }
      }
      
    }
  }
}


R/cpc_read_rawdata.r
#' Read downloaded raw rainfall data from CPC
#'
#' This function reads raw data from CPC, downloaded using `cpc_get_rawdata` and
#' outputs a matrix of values.
#' 
#' Function checks for the existence of the raw files from CPC and only then
#' attempts to process the data. 
#' 
#' The output matrix has 360 rows (latitudes) and 720 columns (longitudes) of 
#' rainfall/precipitation in units of mm/day. The first data point has the lat, 
#' lon values of -89.75 and 0.25 degrees, respectively. Spatial resolution of 
#' the data is 0.5 degrees. Refer to the project home page for plotting the 
#' data with proper North-South orientation. 
#' 
#' By default, output matrix is NOT written to a binary file and raw input files
#' will be DELETED. Change the input function arguments as desired.
#' 
#' @param yr year associated with the downloaded file, 1979 - 2012
#' @param mo month associated with the downloaded file, 1 - 12
#' @param day day associated with the downloaded file, 1 - 28/29/30/31
#' @param remove_input logical, whether or not to remove the raw CPC file 
#' after reading in
#' @param write_output logical, whether or not to write the data to a binary 
#' output file
#' @export
#' @examples
#' # CPC data for Jul 11 2005
#' cpc_1day <- cpc_read_rawdata(2005, 7, 11)
#' # CPC data for Jul 12 2005
#' cpc_1day <- cpc_read_rawdata(2005, 7, 12)

cpc_read_rawdata <- function(yr, mo, day, remove_input = TRUE, write_output = FALSE) {
  
  stopifnot(!(any(c(yr, mo, day) %in% c(""))))
  
  # file name
  dateStr <- paste0(yr, sprintf("%.2d", mo), sprintf("%.2d", day))
  cpcFile <- paste0("raw_", dateStr, ".bin")
  if (yr <= 2008) {
    cpcFile <- paste0("raw_", dateStr, ".gz")
  }
  if(!(file.exists(cpcFile))) {
    stop("Raw file from CPC doesnt exist! First run cpc_get_rawdata()!")
  }
  
  # data attributes, from PRCP_CU_GAUGE_V1.0GLB_0.50deg_README.txt
  cpcNumLat   <- 360 # number of lats
  cpcNumLon   <- 720 # number of lons
  cpcNumBytes <- cpcNumLat * cpcNumLon * 2 # 2 fields, precipitation and num gages
  
  # read data
  # open file connection
  if (yr <= 2008) {
    fileCon <- gzcon(file(cpcFile, "rb"))
  } else {
    fileCon <- file(cpcFile, "rb")
  }
  
  inData  <- readBin(con = fileCon, 
                     what = numeric(), 
                     n = cpcNumBytes, 
                     size = 4)
  close(fileCon)
  # remove input
  if (remove_input) {
    file.remove(cpcFile)
  }
  
  # extract first field (precipitation), ignore second field (num gages)
  inData <- inData[1:(cpcNumBytes/2)]
  
  # CPC raw data goes from South to North; need to flip rows for plotting
  prcpData <- array(0, dim = c(cpcNumLat, cpcNumLon))
  for(eachRow in 1:cpcNumLat) {
    index1 <- 1 + (eachRow - 1) * cpcNumLon
    index2 <- eachRow * cpcNumLon
    prcpData[eachRow, ] <- inData[index1:index2]
  }
  # remove -ve (missing) values
  prcpData[prcpData < 0] <- NA
  # convert tenths of mm to mm
  prcpData <- ifelse(prcpData > 0, prcpData * 0.1, prcpData)
  
  # write data to file
  if (write_output) {
    outCon <- file(paste0(dateStr, ".bin"), "wb")
    writeBin(as.numeric(prcpData), con = outCon, size = 4)
    close(outCon)
  }
  
  return (prcpData)
}
